cellrank/tools/_utils.py

```diff
@@ -1150,3 +1150,156 @@ def _info_if_obs_keys_categorical_present(
+    `(n_samples x n_clusters)` where rows sum to one and indicate cluster membership to each of
+    the `c_clusters` clusters, we compute an assignment of a subset of samples to clusters such
+    that each cluster is represented by its `n_most_likely` most likely samples. In case a sample
+    is assigned more than once, it can either be removed (`remove_overlap = True`) or it can be
+    assigned to the cluster it most likely belongs to (`remove_overlap = False`). In case this
+    leaves a cluster with less than `raise_threshold x n_most_likely` samples, we raise an exception.
+    In case this leaves clusters c_1, ..., c_m with less than `n_most_likely` samples, but more than
+    `raise_threshold x n_most_likely` samples, we append c_1, ..., c_m to a list `critical_clusters`,
+    which we return.
+
+    We return a boolean matrix `a_discrete` of the same shape as `a_fuzzy`, where 1 in position
+    `i, j` indicates that sample `i` is assigned to cluster `j`. Note that we don't assign all samples
+    to clusters (most entries in `a_discrete` will be 0) - this is meant to only assign a small
+    subset of the samples, which we are most confident in.
+
+    Params
+    ------
+    a_fuzzy
+        Numpy array of shape (`n_samples x n_clusters`) representing a fuzzy clustering. Rows must sum to one.
+    n_most_likely
+        Number of samples we want to assign to each cluster.
+    remove_overlap
+        If `True`, remove ambigious samples. Otherwise, assign them to the most likely cluster.
+    raise_threshold
+        If a cluster is assigned less than `raise_threshold x n_most_likely` samples, raise an
+        exception. Set to `None` if you never want to raise an exception, even if this leaves
+        some clusters empty.
+
+    Returns
+    -------
+    a_discrete
+        Boolean matrix of the same shape as `a_fuzzy`, assigning a subset of the samples to clusters.
+    critical_clusters
+        Array of clusters with less than `n_most_likely` samples assigned.
+    """
+
+    # check the inputs
+    if not isinstance(a_fuzzy, np.ndarray):
+        raise TypeError(
+            f"Expected `a_fuzzy` to be of type `numpy.ndarray`, got `{type(a_fuzzy).__name__!r}`."
+        )
+    if not np.allclose(a_fuzzy.sum(1), 1, rtol=1e3 * EPS, atol=1e3 * EPS):
+        raise ValueError("Rows in `a_fuzzy` do not sum to one.")
+
+    # initialise
+    n_clusters = a_fuzzy.shape[1]
+    if raise_threshold is not None:
+        n_raise = int(raise_threshold * n_most_likely)
+
+    # create argmax assignment as a lookup table
+    max_assignment = a_fuzzy.argmax(axis=1)
+
+    # initially select `n_most_likely` samples per cluster
+    sample_assignment = {
+        cl: fuzzy_assignment.argsort()[::-1][:n_most_likely]
+        for cl, fuzzy_assignment in enumerate(a_fuzzy.T)
+    }
+
+    # create the one-hot encoded discrete clustering
+    a_discrete = np.zeros_like(a_fuzzy)
```

Comment:
Should be the dtype something like `int32`?

Reply:
good idea. This matrix will only store 0s and 1s, so could even make it a bool type, no?

Reply:
changed to bool